crates/red_knot_python_semantic/src/types.rs

```diff
@@ -583,7 +583,20 @@ impl<'db> Type<'db> {
             ),
 
             // TODO annotated return type on `__new__` or metaclass `__call__`
-            Type::Class(class) => CallOutcome::callable(Type::Instance(class)),
+            Type::Class(class) => {
+                // If the class is the builtin-bool class (for example `bool(1)`), we try to return
+                // the specific truthiness value of the input arg, `Literal[True]` for the example above.
+                let is_bool = class.is_stdlib_symbol(db, "builtins", "bool");
+                CallOutcome::callable(if is_bool {
+                    arg_types
+                        .first()
+                        .unwrap_or(&Type::Unknown)
+                        .bool(db)
+                        .into_type(db)
+                } else {
+                    Type::Instance(class)
+                })
```

Comment:
This might not fit into the scope, but I've recently added some functions similar to `bool` in #13511, namely `repr` and `str`.

I think there's a pattern where we'll eventually want to extend this piece of code to deal with all currently implemented magic methods (`__bool__`, `__str__`, `__repr__`, ...) that have an associated builtin calling them.

Maybe we could introduce an enumeration for builtins with a method on `ClassType` that returns an `Option<Builtin>` to have something easily extendable like
```rust
impl<'db> ClassType<'db> {
    pub(crate) fn to_maybe_builtin(self, db: &'db dyn Db) -> Option<Builtin> {
        if self.is_stdlib_symbol(db, "builtins", "bool") {
            Some(Builtin::Bool)
        } else if self.is_stdlib_symbol(db, "builtins", "str") {
            Some(Builtin::Str)
        } else if self.is_stdlib_symbol(db, "builtins", "repr") {
            Some(Builtin::Str)
        } else {
            None
        }
    }
 }
```
And in the selected code something along the lines of
```rust
                let maybe_builtin = class.to_maybe_builtin(db);
                CallOutcome::callable(match maybe_builtin {
                    Some(Builtin::Bool) => arg_types
                        .first()
                        .unwrap_or(&Type::Unknown)
                        .bool(db)
                        .into_type(db),
                    Some(Builtin::Str) => arg_types.first().unwrap_or(&Type::Unknown).str(db),
                    Some(Builtin::Repr) => arg_types.first().unwrap_or(&Type::Unknown).repr(db),
                    None => Type::Instance(class),
                })
```
Doing this for `repr` and `str` is obviously outside of scope.

What do you think @carljm ?

Reply:
Yes, I think we will want to do something like this, though I think we may also want to store the `Builtin` enum value in the `ClassType` (and also in `FunctionType` -- this could simplify our support for `reveal_type`) so that we can just call `is_stdlib_symbol` when the class or function is defined, rather than at every call-site.

But I think that can/should be done as a separate PR, rather than as part of this PR; it's easier to get these generalizations right once there are a few cases in place to generalize.